server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -0,0 +1,208 @@
+                    }
+                } else {
+                    coordState.handleJoin(join); // this might fail and bubble up the exception
+                }
+            }
+
+            joinAccumulator.handleJoinRequest(joinRequest.getSourceNode(), joinCallback);
+
+            if (prevElectionWon == false && coordState.electionWon()) {
+                becomeLeader("handleJoin");
+            }
+        }
+    }
+
+    void becomeCandidate(String method) {
+        assert Thread.holdsLock(mutex) : "Legislator mutex not held";
+        logger.debug("{}: becoming CANDIDATE (was {}, lastKnownLeader was [{}])", method, mode, lastKnownLeader);
+
+        if (mode != Mode.CANDIDATE) {
+            mode = Mode.CANDIDATE;
+            joinAccumulator.close(mode);
+            joinAccumulator = joinHelper.new CandidateJoinAccumulator();
+        }
+    }
+
+    void becomeLeader(String method) {
+        assert Thread.holdsLock(mutex) : "Legislator mutex not held";
+        assert mode == Mode.CANDIDATE : "expected candidate but was " + mode;
+        logger.debug("{}: becoming LEADER (was {}, lastKnownLeader was [{}])", method, mode, lastKnownLeader);
+
+        mode = Mode.LEADER;
+        lastKnownLeader = Optional.of(getLocalNode());
+        joinAccumulator.close(mode);
+        joinAccumulator = joinHelper.new LeaderJoinAccumulator();
+    }
+
+    void becomeFollower(String method, DiscoveryNode leaderNode) {
+        assert Thread.holdsLock(mutex) : "Legislator mutex not held";
+        logger.debug("{}: becoming FOLLOWER of [{}] (was {}, lastKnownLeader was [{}])", method, leaderNode, mode, lastKnownLeader);
+
+        if (mode != Mode.FOLLOWER) {
+            mode = Mode.FOLLOWER;
+            joinAccumulator.close(mode);
+            joinAccumulator = new JoinHelper.FollowerJoinAccumulator();
+        }
+
+        lastKnownLeader = Optional.of(leaderNode);
+    }
+
+    // package-visible for testing
+    long getCurrentTerm() {
+        synchronized (mutex) {
+            return coordinationState.get().getCurrentTerm();
+        }
+    }
+
+    // package-visible for testing
+    Mode getMode() {
+        synchronized (mutex) {
+            return mode;
+        }
+    }
+
+    // package-visible for testing
+    DiscoveryNode getLocalNode() {
+        return transportService.getLocalNode();
+    }
+
+    @Override
+    protected void doStart() {
+        CoordinationState.PersistedState persistedState = persistedStateSupplier.get();
+        coordinationState.set(new CoordinationState(settings, getLocalNode(), persistedState));
+    }
+
+    public void startInitialJoin() {
+        synchronized (mutex) {
+            becomeCandidate("startInitialJoin");
+        }
+    }
+
+    @Override
+    protected void doStop() {
+
+    }
+
+    @Override
+    protected void doClose() {
+
+    }
+
+    public void invariant() {
+        synchronized (mutex) {
+            if (mode == Mode.LEADER) {
```

Comment:
We can restore those assertions about the state of the join helper here - i.e. no accumulated joins when leader or follower.